RZAssemblage/Private/RZProxyAssemblage.m

```diff
@@ -0,0 +1,161 @@
+                             forKeyPath:self.keypath
+                                options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld
+                                context:RZProxyKeyPathContext];
+}
+
+- (void)disableObservation
+{
+    [self.representedObject removeObserver:self forKeyPath:self.keypath context:RZProxyKeyPathContext];
+}
+
+- (BOOL)isRepeatingKeyPath
+{
+    // If nextKeyPaths is nil, this is a repeating tree expansion.
+    return self.nextKeyPaths == nil;
+}
+
+- (void)removeObjectFromElementsAtIndex:(NSUInteger)index
+{
+    id object = [self objectInElementsAtIndex:index];
+    RZAssemblageLog(@"%p:Remove %@ at %zd", self, object,  index);
+    [self removeMonitorsForObject:object];
+    [self.childrenStorage removeObjectAtIndex:index];
+}
+
+- (void)insertObject:(NSObject *)object inElementsAtIndex:(NSUInteger)index
+{
+    RZAssemblageLog(@"%p:Insert %@ at %zd", self, object, index);
+    NSParameterAssert(object);
+    [self addMonitorsForObject:object];
+    [self.childrenStorage insertObject:object atIndex:index];
+}
+
+- (id)nodeAtIndex:(NSUInteger)index
+{
+    id node = [super nodeAtIndex:index];
+    if ( node == nil && self.isRepeatingKeyPath ) {
+        id object = [self objectInElementsAtIndex:index];
+        [self removeMonitorsForObject:object];
+        node = [[RZProxyAssemblage alloc] initWithObject:object childKey:self.keypath];
+        [self addMonitorsForObject:node];
+        [self disableObservation];
+        [self.childrenStorage replaceObjectAtIndex:index withObject:node];
+        [self enableObservation];
+    }
+    else if ( node == nil && self.nextKeyPaths.count > 0 ) {
+        id object = [self objectInElementsAtIndex:index];
+        [self removeMonitorsForObject:object];
+        node = [[RZProxyAssemblage alloc] initWithObject:object keypaths:self.nextKeyPaths];
+        [self addMonitorsForObject:node];
+        [self disableObservation];
+        [self.childrenStorage replaceObjectAtIndex:index withObject:node];
+        [self enableObservation];
+    }
+    return node;
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+    if ( context == RZProxyKeyPathContext ) {
+        NSKeyValueChange changeType = [change[NSKeyValueChangeKindKey] unsignedIntegerValue];
+        NSIndexSet *indexes = change[NSKeyValueChangeIndexesKey];
+        [self openBatchUpdate];
+        if ( changeType == NSKeyValueChangeInsertion ) {
+            [indexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
+                [self.changeSet insertAtIndexPath:[NSIndexPath indexPathWithIndex:idx]];
+            }];
+        }
+        else if ( changeType == NSKeyValueChangeRemoval ) {
+            NSArray *oldValues = change[NSKeyValueChangeOldKey];
+            NSUInteger i = 0;
+            NSUInteger index = [indexes firstIndex];
+            for ( ; i < indexes.count; i++ ) {
```

Comment:
I mean... now that this is a for loop you can toss the `NSUInteger i = 0;` in there instead of leaving an empty initialization slot